base_layer/wallet/src/storage/sqlite_db.rs

```diff
@@ -204,33 +229,99 @@ impl WalletSqliteDatabase {
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
+                    CommsSecretKey::from_bytes(decrypted_key_bytes.as_slice())?
+                },
+            };
+
+            Ok(Some(DbValue::CommsSecretKey(secret_key)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_comms_address(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        if let Some(key_str) = WalletSettingSql::get(DbKey::CommsAddress.to_string(), &conn)? {
+            Ok(Some(DbValue::CommsAddress(
+                Multiaddr::from_str(key_str.as_str())
+                    .map_err(|e| WalletStorageError::ConversionError(e.to_string()))?,
+            )))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_comms_features(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        if let Some(key_str) = WalletSettingSql::get(DbKey::CommsFeatures.to_string(), &conn)? {
+            let features = u64::from_str(&key_str).map_err(|e| WalletStorageError::ConversionError(e.to_string()))?;
+            Ok(Some(DbValue::CommsFeatures(features)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn set_tor_id(&self, tor: TorIdentity, conn: &SqliteConnection) -> Result<(), WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        let tor_string = tor
+            .to_json()
+            .map_err(|e| WalletStorageError::ConversionError(e.to_string()))?;
+        match cipher.as_ref() {
+            None => {
+                WalletSettingSql::new(DbKey::TorId.to_string(), tor_string).set(&conn)?;
+            },
+            Some(cipher) => {
+                let ciphertext_integral_nonce = encrypt_bytes_integral_nonce(&cipher, tor_string.into_bytes())
+                    .map_err(|_| WalletStorageError::AeadError("Encryption Error".to_string()))?;
+                WalletSettingSql::new(DbKey::TorId.to_string(), ciphertext_integral_nonce.to_hex()).set(&conn)?;
+            },
+        }
+
+        Ok(())
+    }
+
+    fn get_tor_id(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        if let Some(key_str) = WalletSettingSql::get(DbKey::TorId.to_string(), &conn)? {
+            let tor_json = match cipher.as_ref() {
+                None => key_str,
+                Some(cipher) => {
+                    let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(&key_str)?)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
+                    let tor_json = std::str::from_utf8(&decrypted_key_bytes)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?
+                        .to_string();
+                    // Encryption and decryption breaks the json string, so we fix it.
```

Comment:
This feels weird, the encryption should return the bytes stream byte for byte?

Reply:
Its because we are encoding a string
so this:
"{\"private_key\":{\"Ed25519V3\":\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\"},\"service_id\":\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\",\"onion_port\":18101}"
Becomes:
"\"{\\"private_key\\":{\\"Ed25519V3\\":\\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\\"},\\"service_id\\":\\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\\",\\"onion_port\\":18101}\""

Reply:
Why not serialize it as bytes with serde?

Reply:
I will try and do that when saving as encrypted